ufl/domain.py

```diff
@@ -126,6 +153,119 @@ def _ufl_sort_key_(self):
+    thus the same topological dimension).
+
+    Currently, one can only perform cell integrations when
+    :class:`MixedMesh`es are used.
+
+    .. code-block:: python3
+
+        cell = triangle
+        mesh0 = Mesh(FiniteElement("Lagrange", cell, 1, (2,), identity_pullback, H1))
+        mesh1 = Mesh(FiniteElement("Lagrange", cell, 1, (2,), identity_pullback, H1))
+        domain = MixedMesh([mesh0, mesh1])
+        elem0 = FiniteElement("Lagrange", cell, 1, (), identity_pullback, H1)
+        elem1 = FiniteElement("Lagrange", cell, 2, (), identity_pullback, H1)
+        elem = MixedElement([elem0, elem1])
+        V = FunctionSpace(domain, elem)
+        v = TestFunction(V)
+        v0, v1 = split(v)
+
+    """
+
+    def __init__(self, meshes, ufl_id=None, cargo=None):
+        """Initialise."""
+        self._ufl_id = self._init_ufl_id(ufl_id)
+        # Store reference to object that will not be used by UFL
+        self._ufl_cargo = cargo
+        if cargo is not None and cargo.ufl_id() != self._ufl_id:
+            raise ValueError("Expecting cargo object (e.g. dolfin.Mesh) to have the same ufl_id.")
+        if any(isinstance(m, MixedMesh) for m in meshes):
+            raise NotImplementedError("""
+                Currently component meshes can not include MixedMesh instances""")
+        # currently only support single cell type.
+        (self._ufl_cell,) = set(m.ufl_cell() for m in meshes)
+        (gdim,) = set(m.geometric_dimension() for m in meshes)
+        # TODO: Need to change for more general mixed meshes.
+        (tdim,) = set(m.topological_dimension() for m in meshes)
+        AbstractDomain.__init__(self, tdim, gdim)
+        self._meshes = tuple(meshes)
+
+    def ufl_cargo(self):
+        """Return carried object that will not be used by UFL."""
+        return self._ufl_cargo
+
+    def ufl_cell(self):
+        """Get the cell."""
+        # TODO: Might need MixedCell class for more general mixed meshes.
+        return self._ufl_cell
+
+    def __repr__(self):
+        """Representation."""
+        return "MixedMesh(%s, ufl_id=%s)" % (repr(self._meshes), repr(self._ufl_id))
+
+    def __str__(self):
+        """Format as a string."""
+        return "<MixedMesh #%s>" % (self._ufl_id,)
+
+    def _ufl_hash_data_(self):
+        """UFL hash data."""
+        return ("MixedMesh", self._ufl_id)
+
+    def _ufl_signature_data_(self, renumbering):
+        """UFL signature data."""
+        return ("MixedMesh", tuple(m._ufl_signature_data_(renumbering) for m in self._meshes))
+
+    def _ufl_sort_key_(self):
+        """UFL sort key."""
+        typespecific = (self._ufl_id,)
+        return (self.geometric_dimension(), self.topological_dimension(), "MixedMesh", typespecific)
+
+    @property
+    def meshes(self):
+        """Return the component meshes."""
+        return self._meshes
+
+    def iterable_like(self, element):
+        """Return iterable object that is iterable like ``element``."""
```

Comment:
Assumes that you have `num_sub_elements`, so should be type-hinted.

Reply:
Fixed. `AbstractFiniteElement` declares `num_sub_elements`.